File: api/src/main/java/com/joshb/mcspringboot/util/SpringScanner.java
package com.joshb.mcspringboot.util;

import org.springframework.aop.support.AopUtils;
import org.springframework.context.ApplicationContext;
import org.springframework.stereotype.Component;

import java.lang.annotation.Annotation;
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.Map;

import lombok.AllArgsConstructor;

@Component
@AllArgsConstructor
public class SpringScanner {
    
    private final ApplicationContext context;
    
    public Map<Method, Object> scanMethods(Class<? extends Annotation> annotation) {
        HashMap<Method, Object> methods = new HashMap<>();
        for (String beanName : context.getBeanDefinitionNames()) {
            Object object = context.getBean(beanName);
            
            Class<?> clazz = object.getClass();
            if (AopUtils.isAopProxy(clazz)) {
                clazz = AopUtils.getTargetClass(object);
            }
            
            for (Method method : clazz.getDeclaredMethods()) {
                if (method.isAnnotationPresent(annotation)) {
                    methods.put(method, object);
                }
            }
        }
        return methods;
    }
}

File: mcspring-api/mcspring-base/src/main/kotlin/in/kyle/mcspring/util/SpringScanner.kt
package `in`.kyle.mcspring.util

import org.springframework.aop.support.AopUtils
import org.springframework.context.ApplicationContext
import org.springframework.stereotype.Component
import java.lang.reflect.Method

@Component
class SpringScanner(private val context: ApplicationContext) {

    fun scanMethods(vararg annotations: Class<out Annotation>): Map<Method, Any> {
        val methods = mutableMapOf<Method, Any>()
        for (beanName in context.beanDefinitionNames) {
            val obj = context.getBean(beanName)
            for (method in getRealClass(obj).declaredMethods) {
                if (annotations.any { method.isAnnotationPresent(it) }) {
                    methods[method] = obj
                }
            }
        }
        return methods
    }

    private fun getRealClass(obj: Any): Class<*> {
        return if (AopUtils.isAopProxy(obj.javaClass)) {
            AopUtils.getTargetClass(obj)
        } else {
            obj.javaClass
        }
    }
}
